uniffi_bindgen/src/bindings/backend.rs

```diff
@@ -0,0 +1,80 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use crate::interface::*;
+use std::fmt;
+
+pub type TypeIdentifier = Type;
+pub type Literal = crate::interface::Literal;
+
+// Placeholder return type, while we do refactoring.
+pub type StringReturn = String;
+
+pub trait LanguageOracle {
+    fn find(&self, type_: &TypeIdentifier) -> Result<Box<dyn CodeType>, askama::Error>;
+
+    /// Get the idiomatic Kotlin rendering of a class name (for enums, records, errors, etc).
+    fn class_name(&self, nm: &dyn fmt::Display) -> String;
+
+    /// Get the idiomatic Kotlin rendering of a function name.
+    fn fn_name(&self, nm: &dyn fmt::Display) -> String;
+
+    /// Get the idiomatic Kotlin rendering of a variable name.
+    fn var_name(&self, nm: &dyn fmt::Display) -> String;
+
+    /// Get the idiomatic Kotlin rendering of an individual enum variant.
+    fn enum_variant(&self, nm: &dyn fmt::Display) -> String;
+
+    /// Get the idiomatic Kotlin rendering of an exception name
+    ///
+    /// This replaces "Error" at the end of the name with "Exception".  Rust code typically uses
+    /// "Error" for any type of error but in the Java world, "Error" means a non-recoverable error
+    /// and is distinguished from an "Exception".
+    fn exception_name(&self, nm: &dyn fmt::Display) -> String;
+
+    fn ffi_type_label(&self, ffi_type: &FFIType) -> String;
+}
+
+pub trait CodeType {
```

Comment:
From its presence in a `backend.rs` I infer that this trait may be meant to be shared between the different language backends. That seems valuable, but also it's not immediately obvious to me that it will work out well - for example the Python backend has a "coerce" operation that isn't present in the Kotlin backend. Will we end up with something here that's the union of all possible backend features? Will we end up with a "parent" trait that various backends can specialize?

(That's not intended as stop-energy on this sharing approach, just reacting out loud on initial read-through)

Reply:
Yeah, it's not clear to me yet.

Reply:
Yeah... 
I think if we go with this `CodeType` approach,
>  Will we end up with a "parent" trait that various backends can specialize

or a variation of it resonates more with me more than the alternative. So we end up having specific operations, like `coerce`, not to use this pattern or depend on the `oracle`. Otherwise, if we keep the union of all operations, when the backends are separated, backend authors may want to keep adding to the `CodeType`... and that creates work for **all** the other backends

Could be worth it for us to say "those are the core operations, everything else is backend-specific, and backend authors should write their own logic" 😞